libmamba/src/specs/regex_spec.cpp

```diff
@@ -39,25 +38,60 @@ namespace mamba::specs
     }
 
     RegexSpec::RegexSpec()
-        : RegexSpec(std::regex(free_pattern.data(), free_pattern.size()), std::string(free_pattern))
+        : RegexSpec(std::string(free_pattern))
     {
     }
 
-    RegexSpec::RegexSpec(std::regex pattern, std::string raw_pattern)
-        : m_pattern(std::move(pattern))
-        , m_raw_pattern(std::move(raw_pattern))
+    RegexSpec::RegexSpec(std::string raw_pattern)
     {
+        // If the string is wrapped in `^` and `$`, `conda.model.MatchSpec` considers it a regex.
+        // See:
+        // https://github.com/conda/conda/blob/52b6393d6331e8aa36b2e23ab65766a980f381d2/conda/models/match_spec.py#L134-L139.
+        // See:
+        // https://github.com/conda/conda/blob/52b6393d6331e8aa36b2e23ab65766a980f381d2/conda/models/match_spec.py#L889-L894
+        if (util::starts_with(raw_pattern, pattern_start) && util::ends_with(raw_pattern, pattern_end))
+        {
+            m_raw_pattern = raw_pattern;
+            m_pattern = std::regex(m_raw_pattern);
+            return;
+        }
```

Comment:
I would find it clearer to have something like:
```cpp
RegexSpec::RegexSpec(std::string raw_pattern)
    : m_raw_pattern(format_raw_pattern(std::move(raw_pattern))
    , m_pattern(std::regex(m_raw_pattern))
{
}
```

And have all the logic in `format_raw_pattern` (or any better name you can come with) rather than in the constructor, with a return statement.

Reply:
I thought of this, but does this guarantee that `m_raw_pattern` is strictly initialized before `m_pattern`?

I am not sure it is safe as `m_raw_pattern` is declared before `m_pattern`:

https://github.com/mamba-org/mamba/blob/ad9b2d6b8840825411a3c23251cfbb978ab60e89/libmamba/include/mamba/specs/regex_spec.hpp#L64-L65

Reply:
Ah good catch, you have to invert them in the declaration. The intialization order is guanranteed to be the same as the declaration order/

Reply:
Could this change the ABI?